src/mame/barcrest/mpu4_oki_sampled_sound.cpp

```diff
@@ -0,0 +1,199 @@
+
+/*
+
+Universal sampled sound program card PCB 683077
+Sampled sound card, using a PIA and PTM for timing and data handling
+The sound hardware is technically twin channel stereo, but there's no evidence to suggest it was ever connected as such
+I've separated the channels here, to tie back at the game level
+
+*/
+
+#include "emu.h"
+
+#include "mpu4_oki_sampled_sound.h"
+
+DEFINE_DEVICE_TYPE(MPU4_OKI_SAMPLED_SOUND, mpu4_oki_sampled_sound, "mpu4okisnd", "Barcrest Sampled Sound Program Card")
+
+mpu4_oki_sampled_sound::mpu4_oki_sampled_sound(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock)
+	: device_t(mconfig, MPU4_OKI_SAMPLED_SOUND, tag, owner, clock),
+		device_mixer_interface(mconfig, *this),
+		m_cb2_handler(*this),
+		m_ptm_ic3ss(*this, "ptm_ic3ss"),
+		m_pia_ic4ss(*this, "pia_ic4ss"),
+		m_msm6376(*this, "msm6376")
+{
+}
+
+/*
+The MSM6376 sound chip is configured in a slightly strange way, to enable dynamic
+sample rate changes (8Khz, 10.6 Khz, 16 KHz) by varying the clock.
+According to the BwB programmer's guide, the formula is:
+MSM6376 clock frequency:-
+freq = (CPU_CLOCK/((t3L+1)(t3H+1)))*[(t3H(T3L+1)+1)/(2(t1+1))]
+where [] means rounded up integer,
+t3L is the LSB of Clock 3,
+t3H is the MSB of Clock 3,
+and t1 is the initial value in clock 1.
+
+As originally made, this is the peripheral clock (1.72 MHz), but for video it's the E Clock of the 68k (1MHz)
+
+//O3 -> G1  O1 -> c2 o2 -> c1
+
+This is a bit of a cheat - since we don't clock into the OKI chip directly, we need to
+calculate the oscillation frequency in advance. We're running the timer for interrupt
+purposes, but the frequency calculation is done by plucking the values out as they are written.
+*/
+
+void mpu4_oki_sampled_sound::ic3_write(offs_t offset, uint8_t data)
+{
+	m_ptm_ic3ss->write(offset, data);
+
+	if (offset == 3)
+	{
+		m_t1 = data;
+	}
+	if (offset == 6)
+	{
+		m_t3h = data;
+	}
+	if (offset == 7)
+	{
+		m_t3l = data;
+	}
+
+	float const num = float(clock()) / ((m_t3l + 1) * (m_t3h + 1));
+	float const denom = std::ceil(float(m_t3h * (m_t3l + 1) + 1) / (2 * (m_t1 + 1))); //need to round up, this gives same precision as chip
+
+	int const freq = int(num * denom);
+
+	if (freq)
+	{
+		m_msm6376->set_unscaled_clock(freq);
+	}
+}
+
+uint8_t mpu4_oki_sampled_sound::ic3_read(offs_t offset)
+{
+	return m_ptm_ic3ss->read(offset);
+}
+
+void mpu4_oki_sampled_sound::ic4_write(offs_t offset, uint8_t data)
+{
+	m_pia_ic4ss->write(offset, data);
+}
+
+uint8_t mpu4_oki_sampled_sound::ic4_read(offs_t offset)
+{
+	return m_pia_ic4ss->read(offset);
+}
+
+void mpu4_oki_sampled_sound::pia_gb_porta_w(uint8_t data)
+{
+	LOG_SS(("%s: GAMEBOARD: PIA Port A Set to %2x\n", machine().describe_context(), data));
+	m_msm6376->write(data);
+}
+
+void mpu4_oki_sampled_sound::pia_gb_portb_w(uint8_t data)
+{
+	uint8_t changed = m_expansion_latch^data;
+
+	LOG_SS(("%s: GAMEBOARD: PIA Port B Set to %2x\n", machine().describe_context(), data));
+
```

Comment:
Since this device is self-contained now, please use `LOG` from `logmacro.h` for configurable logging in this file.